### PR TITLE
docs: rewrite inputs/outputs section of Template Syntax and add example

### PR DESCRIPTION
**Pull request by kapunahelewong**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [x] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## What is the new behavior?
Complete rewrite of the Inputs/Outputs section of Template Syntax along with an example that demonstrates the principles of the section.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Other information

This is the 9th in a series of PRs that together are a complete rewrite of Template Syntax. Once they are all merged, the next step is to break them into individual docs as in the NgModules section.

This particular section focuses on Inputs/Outputs with a complete rewrite. It also includes an example demonstrating all the points covered in the text. There are also written explanations of the new diagrams (svgs created in Angular colors) in case of devs using assistive technology.

### CODE REVIEW

File: aio/content/guide/template-syntax.md [hunk 1]
```diff
@@ -1658,171 +1658,444 @@ This example declares the `fax` variable as `ref-fax` instead of `#fax`.
 
 {@a inputs-outputs}
 
-## Input and Output properties
+## `@Input()` and `@Output()` properties
 
-An _Input_ property is a _settable_ property annotated with an `@Input` decorator.
-Values flow _into_ the property when it is data bound with a [property binding](#property-binding)
+`@Input()` and `@Output()` let parent and child components share data with one
+another. The `@Input()` property is settable while the `@Output()` property is observable.
```
Comment (trotyl) [inline aio/content/guide/template-syntax.md]: Would `writable` better than `settable` to align with ECMAScript term?

File: aio/content/guide/template-syntax.md [hunk 2]
```diff
@@ -1658,171 +1658,444 @@ This example declares the `fax` variable as `ref-fax` instead of `#fax`.
...
 
-**You can always bind to a public property of a component in its own template.**
-It doesn't have to be an _Input_ or _Output_ property
 
-A component's class and template are closely coupled.
-They are both parts of the same thing.
-Together they _are_ the component.
-Exchanges between a component class and its template are internal implementation details.
+<figure>
+  <img src='generated/images/guide/inputs-outputs/input.gif' alt="Input data flow diagram">
```
Comment (trotyl) [inline aio/content/guide/template-syntax.md]: Another quote style issue as well as other `<figure>`s (but less important as not exposed to user)

File: aio/content/guide/template-syntax.md [hunk 3]
```diff
@@ -1658,171 +1658,444 @@ This example declares the `fax` variable as `ref-fax` instead of `#fax`.
...
+`@Output()` almost always returns an Angular [`EventEmitter`](api/core/EventEmitter) with values flowing out of the component as [events](#event-binding).
+
+
+<figure>
+  <img src='generated/images/guide/inputs-outputs/output.gif' alt="Output diagram">
+</figure>
+
+Just like with `@Input()`, you can use `@Output()`
+on any property of the child component but its type must be
+an event.
```
Comment (trotyl) [inline aio/content/guide/template-syntax.md]: `event` -> ``` `EventEmitter` ```?

File: aio/content/guide/template-syntax.md [hunk 4]
```diff
@@ -1658,171 +1658,444 @@ This example declares the `fax` variable as `ref-fax` instead of `#fax`.
...
+
+Next, create an `addNewItem()` method in the same component class:
+
+```ts
+export class ItemDetailComponent {
+  @Output() newItemEvent: EventEmitter<string> = new EventEmitter<string>(); //same as above
+
+  addNewItem(value:string) {
+    this.newItemEvent.emit(value);
+}
```
Comment (trotyl) [inline aio/content/guide/template-syntax.md]: Seems missing an end curly brace here

File: aio/content/examples/inputs-outputs/src/app/app.component.html [hunk 5]
```diff
@@ -0,0 +1,40 @@
+
+<app-item-detail [item]="currentItem"></app-item-detail>
+
+<hr>
+
+<app-item-output (newItemEvent)='addItem($event)'></app-item-output>
```
[the remark this answers is not in the document]
Comment (kapunahelewong) [reply aio/content/examples/inputs-outputs/src/app/app.component.html]: Perfect. Done!

File: aio/content/guide/template-syntax.md [hunk 6]
```diff
@@ -1658,171 +1658,444 @@ This example declares the `fax` variable as `ref-fax` instead of `#fax`.
 
 {@a inputs-outputs}
 
-## Input and Output properties
+## `@Input()` and `@Output()` properties
 
-An _Input_ property is a _settable_ property annotated with an `@Input` decorator.
-Values flow _into_ the property when it is data bound with a [property binding](#property-binding)
+`@Input()` and `@Output()` let parent and child components share data with one
```
[the remark this answers is not in the document]
Comment (kapunahelewong) [reply aio/content/guide/template-syntax.md]: Changed. :)

File: aio/content/guide/template-syntax.md (hunk 1 above)
Comment (kapunahelewong) [reply aio/content/guide/template-syntax.md]: Works for me. Done!

File: aio/content/guide/template-syntax.md (hunk 2 above)
Comment (kapunahelewong) [reply aio/content/guide/template-syntax.md]: ah thanks for catching these. I think I got them all.

File: aio/content/guide/template-syntax.md (hunk 4 above)
Comment (kapunahelewong) [reply aio/content/guide/template-syntax.md]: Thanks for catching. It's just placeholder though as I have to add the docregions (the real code snippets from the example). Fixed it anyway as a reminder to myself.

File: aio/content/guide/template-syntax.md (hunk 3 above)
Comment (kapunahelewong) [reply aio/content/guide/template-syntax.md]: Ah yes, I like that better too. Thanks!

File: aio/content/guide/template-syntax.md [hunk 7]
```diff
@@ -1672,170 +1672,360 @@ This example declares the `fax` variable as `ref-fax` instead of `#fax`.
...
 
-<div class="alert is-helpful">
+The different parts of the above declaration are as follows:
 
-Alternatively, you can identify members in the `inputs` and `outputs` arrays
-of the directive metadata, as in this example:
+* `@Output()`&mdash;a decorator function marking the property as a way for data to go from the child to the parent
+* `newItemEvent`&mdash;the name of the `@Output()`
+* `EventEmitter<string>`&mdash;the `@Output()`'s type
+* `new EventEmitter<string>()`&mdash;tells Angular to create a new event and that the data it emits is of type string. The type could be any type, such as `number`, `boolean`, and so on. For more information on `EventEmitter`, see the [EventEmitter API documentation](api/core/EventEmitter).
```
Comment (trotyl) [inline aio/content/guide/template-syntax.md]:
```suggestion
* `new EventEmitter<string>()`&mdash;tells Angular to create a new event emitter and that the data it emits are of type string. The type could be any type, such as `number`, `boolean`, and so on. For more information on `EventEmitter`, see the [EventEmitter API documentation](api/core/EventEmitter).
```

File: aio/content/guide/template-syntax.md [hunk 8]
```diff
@@ -1672,170 +1672,360 @@ This example declares the `fax` variable as `ref-fax` instead of `#fax`.
...
 </code-example>
 
-<div class="alert is-helpful">
+The different parts of the above declaration are as follows:
 
-Alternatively, you can identify members in the `inputs` and `outputs` arrays
-of the directive metadata, as in this example:
+* `@Output()`&mdash;a decorator function marking the property as a way for data to go from the child to the parent
+* `newItemEvent`&mdash;the name of the `@Output()`
+* `EventEmitter<string>`&mdash;the `@Output()`'s type
```
Comment (trotyl) [inline aio/content/guide/template-syntax.md]: Note, there's a [`no-inferrable-types`](https://palantir.github.io/tslint/rules/no-inferrable-types/) rule that forbids duplication in type signature and initializer, which is [by default enabled](https://github.com/angular/angular-cli/blob/30345c34995e1d683f6b96395635ecaf749cfbb1/packages/schematics/angular/workspace/files/tslint.json.template#L44-L47) in Angular CLI. I'm not sure whether doing this is right in example.

File: aio/content/guide/template-syntax.md [hunk 9]
```diff
@@ -1672,170 +1672,360 @@ This example declares the `fax` variable as `ref-fax` instead of `#fax`.
...
-</div>
+The `addNewItem()` function uses the `@Output()`, `newItemEvent`,
+to raise an event in which it emits the value the user
+types into the `<input>`. In other words, when
+the user clicks the add button in the UI, the child lets the parent know
+about the event and gives that data to the parent.
+
+#### In the child's template
+
+The child's template has two elements. The first is an HTML `<input>` with a
```
Comment (trotyl) [inline aio/content/guide/template-syntax.md]: The `<input>` is wrapped into a `<label>`, which is mathematically three elements. The most simple change may be `two controls`?

File: aio/content/guide/template-syntax.md [hunk 10]
```diff
@@ -1672,170 +1672,360 @@ This example declares the `fax` variable as `ref-fax` instead of `#fax`.
 
 {@a inputs-outputs}
 
-## Input and Output properties
+## `@Input()` and `@Output()` properties
 
-An _Input_ property is a _settable_ property annotated with an `@Input` decorator.
-Values flow _into_ the property when it is data bound with a [property binding](#property-binding)
+`@Input()` and `@Output()` allow Angular to share data between the parent context and child directives. The `@Input()` property is writable while the `@Output()` property is observable.
```
[the remark this answers is not in the document]
Comment (kapunahelewong) [reply aio/content/guide/template-syntax.md]: Fixed.

File: aio/content/guide/template-syntax.md [hunk 11]
```diff
@@ -1672,170 +1672,360 @@ This example declares the `fax` variable as `ref-fax` instead of `#fax`.
...
+It helps to remember that the data flow is from the perspective of the
+child component. So an `@Input()` allows data to be input _into_ the
+child component from the parent component.
+
+
+<figure>
+  <img src="generated/images/guide/inputs-outputs/input.svg" alt="Input data flow diagram">
+</figure>
+
+To use `@Input()`, you need to edit these parts of your app:
```
[the remark this answers is not in the document]
Comment (kapunahelewong) [reply aio/content/guide/template-syntax.md]: Ah ok. Updated!

File: aio/content/guide/template-syntax.md [hunk 12]
```diff
@@ -1672,170 +1672,360 @@ This example declares the `fax` variable as `ref-fax` instead of `#fax`.
...
+<code-example path="inputs-outputs/src/app/item-detail/item-detail.component.ts" region="use-input" header="src/app/item-detail/item-detail.component.ts" linenums="false">
 </code-example>
 
-The `iconUrl` and `onSave` are members of the `AppComponent` class.
-They are _not_ decorated with `@Input()` or `@Output`.
-Angular does not object.
 
-**You can always bind to a public property of a component in its own template.**
-It doesn't have to be an _Input_ or _Output_ property
+In this case, `@Input()` decorates the property `item`, which has
```
[the remark this answers is not in the document]
Comment (kapunahelewong) [reply aio/content/guide/template-syntax.md]: @gkalpak do you know if there's something I can do to keep `item` from turning into a link?

File: aio/content/guide/template-syntax.md [hunk 13]
```diff
@@ -1672,170 +1672,360 @@ This example declares the `fax` variable as `ref-fax` instead of `#fax`.
...
-The component author is in complete control of those bindings.
+The target in the square brackets, `[]`, is the property you decorate
+with `@Input()` in the child component. The binding source, the part
+to the right of the equal sign, is the data that the parent
+component passes to the nested component.
 
-But other components shouldn't have that kind of unrestricted access.
-You'd have a hard time supporting your component if anyone could bind to any of its properties.
-Outside components should only be able to bind to the component's public binding API.
+The key takeaway is that to use a child component's property in a binding in a parent component&mdash;that is, what's
```
[the remark this answers is not in the document]
Comment (kapunahelewong) [reply aio/content/guide/template-syntax.md]: This is much better. Thank you!

File: aio/content/guide/template-syntax.md [hunk 14]
```diff
@@ -1672,170 +1672,360 @@ This example declares the `fax` variable as `ref-fax` instead of `#fax`.
...
-The following discussion refers to _components_ for brevity and
-because this topic is mostly a concern for component authors.
 </div>
 
-<h3 class="no-toc">Discussion</h3>
+{@a input}
+
+## How to use `@Input()`
+
+Use the `@Input()` decorator in a child component to let Angular know
```
[the remark this answers is not in the document]
Comment (kapunahelewong) [reply aio/content/guide/template-syntax.md]: Got it. Changed! :)

File: aio/content/guide/template-syntax.md [hunk 15]
```diff
@@ -1672,170 +1672,360 @@ This example declares the `fax` variable as `ref-fax` instead of `#fax`.
...
+
+{@a output}
 
-#### TypeScript _public_ doesn't matter
+## How to use `@Output()`
+
+Use the `@Output()` decorator in the child component to allow data to flow from
+the child _out_ to the parent.
+
+`@Output()` almost always returns an Angular [`EventEmitter`](api/core/EventEmitter) with values flowing out of the component as [events](#event-binding).
```
[the remark this answers is not in the document]
Comment (kapunahelewong) [reply aio/content/guide/template-syntax.md]: Got it. Done.

File: aio/content/guide/template-syntax.md [hunk 16]
```diff
@@ -1672,170 +1672,360 @@ This example declares the `fax` variable as `ref-fax` instead of `#fax`.
...
+
+<figure>
+  <img src="generated/images/guide/inputs-outputs/output.svg" alt="Output diagram">
+</figure>
 
-You can't use the TypeScript _public_ and _private_ access modifiers to
-shape the component's public binding API.
+Just like with `@Input()`, you can use `@Output()`
+on any property of the child component but its type must be
+`EventEmitter`.
```
[the remark this answers is not in the document]
Comment (kapunahelewong) [reply aio/content/guide/template-syntax.md]: Changed "must" to "should". It gives the leeway you point out but sticks to the intention of the guide. Thanks!

File: aio/content/guide/template-syntax.md [hunk 17]
```diff
@@ -1672,170 +1672,360 @@ This example declares the `fax` variable as `ref-fax` instead of `#fax`.
...
+
+
+<figure>
+  <img src="generated/images/guide/inputs-outputs/output.svg" alt="Output diagram">
+</figure>
 
-You can't use the TypeScript _public_ and _private_ access modifiers to
-shape the component's public binding API.
+Just like with `@Input()`, you can use `@Output()`
+on any property of the child component but its type must be
```
[the remark this answers is not in the document]
Comment (kapunahelewong) [reply aio/content/guide/template-syntax.md]: Got it. Fixed.

File: aio/content/guide/template-syntax.md [hunk 18]
```diff
@@ -1672,170 +1672,360 @@ This example declares the `fax` variable as `ref-fax` instead of `#fax`.
...
+through which data can travel from the child to the parent.
+The child component then has to raise an event so the
+parent knows something has changed. To raise an event,
+`@Output()` works hand in hand with `EventEmitter`,
+which is a class in `@angular/core` that you
+use to emit custom events.
 
-All data bound properties must be TypeScript _public_ properties.
-Angular never binds to a TypeScript _private_ property.
+When you use `@Output()`, be sure to edit these parts of your app:
```
[the remark this answers is not in the document]
Comment (kapunahelewong) [reply aio/content/guide/template-syntax.md]: Done!

File: aio/content/guide/template-syntax.md (hunk 7 above)
Comment (kapunahelewong) [reply aio/content/guide/template-syntax.md]: Changed.

File: aio/content/guide/template-syntax.md (hunk 8 above)
Comment (kapunahelewong) [reply aio/content/guide/template-syntax.md]: Hm, @trotyl what should I change here for accuracy?

File: aio/content/guide/template-syntax.md (hunk 9 above)
Comment (kapunahelewong) [reply aio/content/guide/template-syntax.md]: Fixed.

File: aio/content/guide/template-syntax.md (hunk 8 above)
Comment (trotyl) [reply aio/content/guide/template-syntax.md]: `@Output() newItemEvent: EventEmitter<string> = new EventEmitter<string>()` will violating this rule, should be `@Output() newItemEvent = new EventEmitter<string>()` where the type is automatically inferred.

<sub>It's hard to say violating a lint rule is something incorrect, I think it should be consistent across all examples and up to doc team to determine.</sub>

File: aio/content/guide/template-syntax.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (kapunahelewong) [reply aio/content/guide/template-syntax.md]: In an offline conversation, @gkalpak showed me this: 

`<code class="no-auto-link">item</code>`

And it works perfectly.

Filed issue and closed with solution: #30975

File: aio/content/guide/template-syntax.md (hunk 8 above)
Comment (kapunahelewong) [reply aio/content/guide/template-syntax.md]: Got it! Thanks :)

**Review by brandonroberts — APPROVED** [review]

**Comment by brandonroberts** [comment]
Caretaker note: The Codefresh job is unrelated and still pending. I am using a global approval

**Comment by AndrewKushnir** [comment]
@kapunahelewong, it looks like this PR now has a conflict with master branch after merging PR 25561 that also updated .github/CODEOWNERS file. Could you please rebase this PR and resolve conflict(s)? Feel free to add "merge" label back once PR is updated. Thank you.

**Pull request closed without merging**